src/errors.ts

```diff
@@ -0,0 +1,58 @@
+export class BaseError {
+  constructor(private message: string, private data?: unknown) {}
```

Comment:
I would not put an unknown `data` property, just add new props in subclasses. Since these errors will not be serialized anyway it's cleaner.